source/Irrlicht/CImageLoaderBMP.cpp

```diff
@@ -45,20 +45,25 @@ bool CImageLoaderBMP::isALoadableFileFormat(io::IReadFile* file) const
 	return headerID == 0x4d42;
 }
 
+// check whether &p[0] to &p[_off - 1] can be accessed
+#define CHECKP(_off) if ((_off) < 0 || p + (_off) > pEnd) goto exit
+// same for d
+#define CHECKD(_off) if ((_off) < 0 || d + (_off) > destEnd) goto exit

```

Comment:
Technically computing `d + _off` is UB if it is actually out of bounds. And I’ve heard of compilers that elide such a check on this basis.

Reply:
I tried to google this but didn't find anything. Any references?

Reply:
https://en.cppreference.com/w/cpp/language/pointer
>Indirection through an invalid pointer value and passing an invalid pointer value to a deallocation function have undefined behavior. Any other use of an invalid pointer value has implementation-defined behavior. Some implementations might define that copying an invalid pointer value causes a system-generated runtime fault.

Reply:
https://en.cppreference.com/w/c/language/operator_arithmetic
> The behavior is defined only if both the original pointer and the result pointer are pointing at elements of the same array or one past the end of that array.

Or to be pedantic, [expr.add].4:
> When an expression that has integral type is added to or subtracted from a pointer, the result has the type
of the pointer operand. If the expression $P$ points to element $x[i]$ of an array object $x$ with $n$ elements,
the expressions $P + J$ and $J + P$ (where $J$ has the value $j$) point to the (possibly-hypothetical) element
$x[i + j]$ if $0 ≤ i + j ≤ n$; **otherwise, the behavior is undefined.** Likewise, the expression $P - J$ points to the
(possibly-hypothetical) element $x[i − j]$ if $0 ≤ i − j ≤ n$; otherwise, the behavior is undefined.

(emphasis mine)

Reply:
Fixed using `uintptr_t` casts